### ISSUE TITLE
User defined IngressGateway not configured correctly with istioctl

### ISSUE THREAD

**Issue by rytswd** [istio/istio#23303]
**Bug description**

A user defined IngressGateway configuration is not handled correctly, and results in incorrect configuration, and some duplication.

The resources that are incorrectly handled are:
- HorizontalPodAutoscaler
- Deployment
- Gateway
- Service

These get the default `metadata.labels` and `spec.selector` regardless of the user defined configuration, and in case of Deployment, it also uses the default `spec.template.metadata.labels` except for `service.istio.io/canonical-name`.

Also, the following resources are completely duplicated:
- PodDisruptionBudget
- Role
- RoleBinding
- ServiceAccount

There seems to be some fix provided in v1.5.2 for #22703 following #21989. Also, #21996 seems to touch on the similar issue. But not all of the problems seem to be handled.

**Expected behavior**

Two ingress gateways would get dedicated resources, allowing segregated control.

**Steps to reproduce the bug**

Use the following definition as `istioctl manifest generate -f` input

```yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio-customised
spec:
  components:
    ingressGateways:
      - name: istio-ingressgateway
        enabled: true
        k8s:
          hpaSpec:
            minReplicas: 2
            maxReplicas: 3
            metrics:
              - resource:
                  name: cpu
                  targetAverageUtilization: 80
                type: Resource
            scaleTargetRef:
              apiVersion: apps/v1
              kind: Deployment
              name: istio-ingressgateway
      - name: custom-ingressgateway
        namespace: istio-system
        enabled: true
        label:
          app: custom-gateway
          custom: ingressgateway
        k8s:
          hpaSpec:
            minReplicas: 3
            maxReplicas: 4
            metrics:
              - resource:
                  name: cpu
                  targetAverageUtilization: 80
                type: Resource
            scaleTargetRef:
              apiVersion: apps/v1
              kind: Deployment
              name: custom-ingressgateway
```

This then generates resources like

```yaml
apiVersion: autoscaling/v2beta1
kind: HorizontalPodAutoscaler
metadata:
  labels:
    app: istio-ingressgateway
    istio: ingressgateway
    release: istio
  name: custom-ingressgateway
  namespace: istio-system
spec:
  maxReplicas: 4
  metrics:
  - resource:
      name: cpu
      targetAverageUtilization: 80
    type: Resource
  minReplicas: 3
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: custom-ingressgateway
​
---
```

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: custom-ingressgateway
  namespace: istio-system
  labels:
    app: istio-ingressgateway
    istio: ingressgateway
    
    release: istio
spec:
  selector:
    matchLabels:
      app: istio-ingressgateway
      istio: ingressgateway
      
  strategy:
    rollingUpdate:
      maxSurge: 100%
      maxUnavailable: 25%
  template:
    metadata:
      labels:
        app: istio-ingressgateway
        istio: ingressgateway
        
        heritage: Tiller
        release: istio
        chart: gateways
        service.istio.io/canonical-name: custom-ingressgateway
        service.istio.io/canonical-revision: "1.5"
      annotations:
        sidecar.istio.io/inject: "false"
    spec:
      serviceAccountName: istio-ingressgateway-service-account
      # ...
```

You can see only few fields get the custom name `custom-ingressgateway`.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

1.5.2

**Comment by lizongshen** [comment]
Any progress? I have the same problem

**Comment by rytswd** [comment]
I can see a few other issues such as #23341 and #23392 have been newly filed since this issue was raised. Also, [some workaround](https://github.com/istio/istio/issues/19263#issuecomment-609094185) is mentioned - it is certainly difficult to manage, and probably won't solve the issue of having duplicated resource definitions.

I am not familiarized enough with IstioOperator resource handling, but both `istioctl` and `istio-operator` seem to use the same configuration under `manifests/` directory as mentioned in https://github.com/istio/istio/issues/23317#issuecomment-620294455.

Multiple IngressGateway support (and I assume the same can be said to EgressGateway) seems to be having a troubling Helm chart setup under https://github.com/istio/istio/tree/master/manifests/charts/gateways/istio-ingress/templates, where it assumes the setup is done for single IngressGateway with predefined values.

The proper solution would be to make sure user-defined values are handled correctly in the template. I may be able to look into it myself, but it could take some time to get a PR out.

**Comment by sofam** [comment]
I am also hit by this:

```
k describe deployment -n istio-ingress-internal internal-ingressgateway 
Name:                   internal-ingressgateway
Namespace:              istio-ingress-internal
CreationTimestamp:      Mon, 04 May 2020 19:44:24 +0200
Labels:                 app=istio-ingressgateway
                        install.operator.istio.io/owner-group=install.istio.io
                        install.operator.istio.io/owner-kind=IstioOperator
                        install.operator.istio.io/owner-name=istio-controlplane
                        istio=ingressgateway
                        release=istio
Annotations:            deployment.kubernetes.io/revision: 1
                        install.operator.istio.io/chart-owner: IngressGateways
Selector:               app=istio-ingressgateway,istio=ingressgateway
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 100% max surge
Pod Template:
  Labels:           app=istio-ingressgateway
                    istio=ingressgateway
                    service.istio.io/canonical-name=internal-ingressgateway
                    service.istio.io/canonical-revision=1.5
  Annotations:      sidecar.istio.io/inject: false
  Service Account:  istio-ingressgateway-service-account
```

even though the operator YAML looks like this:
```
      - enabled: true
        name: internal-ingressgateway
        namespace: istio-ingress-internal
        label:
          app: internal-ingressgateway
          istio: http2-ingress
```

**Comment by ostromart** [comment]
This is fixed in 1.6 but I'm not sure what it would take to backport to 1.5: https://github.com/istio/istio/pull/22981

**Comment by rytswd** [comment]
Thanks for the pointer @ostromart, it actually looks like it's already backported to v1.5 with #23026.

For my own learning and future contribution, would you be able to shed some light on how `istioctl` handles the IstioOperator spec? I had thought the rendering internally relies on Helm templating, and I was checking out templates under `manifests/` such as 
https://github.com/istio/istio/blob/release-1.5/manifests/gateways/istio-ingress/templates/deployment.yaml. I think I'm mistaken, as that would raise a few more setup questions.

**Comment by ostromart** [comment]
You're right, I noticed that backport just now myself. Under the hood, it's a combination of helm template library to maintain compatibility with the values.yaml API and helm charts in general, plus some overlay code for the other part of the IstioOperator API.

**Comment by rytswd** [comment]
Appreciate your input, I could confirm how Namespace handling is not as per Helm chart but is handled correctly with v1.5.2 already. I'll look to familiarise myself more around the actual code.

As to the above fix, I have built `istioctl` off of `master` branch and tested. It is certainly handling multiple gateways better, but have some questions:

### Duplicated Resources

Purely duplicated resources are:
- Role
- RoleBinidng

I suppose both would simply need `metadata.name` to align with other files: `{{ $gateway.name | default "istio-ingressgateway" }}-sds` instead of hard-coding to be `istio-ingressgateway-sds`.

- ServiceAccount `istio-ingressgateway-service-account`

The custom label is handled, but the resource name conflicts.

The fix seems easy enough on Helm chart for these, so I created a PR #23570.

### Duplicated Resources when `values.global.multiCluster.enabled`

When enabling multicluster, the below resources get duplicated, even with the latest code:
- Gateway `istio-multicluster-ingressgateway`
- EnvoyFilter `istio-multicluster-ingressgateway`
- DestinationRule `istio-multicluster-ingressgateway`

This means the label setup would conflict, like the ServiceAccount above.
`istio-ingressgateway` would have:
```yaml
labels:
  app: istio-ingressgateway
  istio: ingressgateway
```

Whereas `custom-ingressgateway` would have:
```yaml
labels:
  app: custom-ingressgateway
  istio: ingressgateway
  custom: ingressgateway
```

Is it safe to follow the same pattern and take in the gateway name as the resource name?

### Duplicated Resources when `values.global.meshExpansion.enabled`

When enabling mesh expansion, the below resources are duplicated:

- Gateway `meshexpansion-gateway`
- VirtualService `meshexpansion-vs-istiod`
- DestinationRule `meshexpansion-dr-istiod`

The Gateway resource above points to the default selector of `istio: ingressgateway` only - I am guessing this one also needs similar handling, but would appreciate some guidance on this.

**Comment by ostromart** [comment]
I think we need to decide whether we want separate sets of resources for each GW or have them combined. If the latter, duplication is ok as long as it's consistent - apply will just write the same resource twice and no one will be the wiser. Thoughts @howardjohn ?

**Comment by rytswd** [comment]
This was discussed in the #security channel in Slack. @costinm and @nrjpoddar kindly stepped in to share their view of separate SAs being the preference as of now. If no objection raised in the coming few days, I think I can work on backporting the same change in v1.5 and v1.6 sometime next week.

**Comment by marshallford** [comment]
I think I'm still seeing this behavior in 1.6.

```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istiocontrolplane
  namespace: istio-system
spec:
  profile: minimal
  components:
    ingressGateways:
      - name: external-ingressgateway
        enabled: true
        label:
          app: external-ingressgateway
```

Service:
```
❯ k describe svc -n istio-system external-ingressgateway 
Name:                     external-ingressgateway
Namespace:                istio-system
Labels:                   app=external-ingressgateway
                          install.operator.istio.io/owning-resource=istiocontrolplane
                          istio=ingressgateway
```

Deployment (with default `app` label)
```
❯ k describe pods -n istio-system external-ingressgateway-567c7b4874-j4k8h 
Name:         external-ingressgateway-567c7b4874-j4k8h
Namespace:    istio-system
Priority:     0
Node:         ubuntu.lan/10.0.0.5
Start Time:   Sat, 23 May 2020 16:16:29 -0500
Labels:       app=istio-ingressgateway
              chart=gateways
              heritage=Tiller
              istio=ingressgateway
              pod-template-hash=567c7b4874
              release=istio
              service.istio.io/canonical-name=external-ingressgateway
              service.istio.io/canonical-revision=latest
```

**Comment by rytswd** [comment]
@marshallford 
I'm not seeing the same behaviour when using `istioctl`. Both `istioctl manifest apply` and `istioctl manifest generate` report that the Deployment has the correct labels associated for me. How are you installing Istio with that IstioOperator CR? Provided that the Service is getting the label associated correctly, would you be able to double check what you see in the YAML definitions?

`kubectl get service -n istio-system external-ingressgateway -o yaml`
`kubectl get deployment -n istio-system external-ingressgateway -o yaml`

**Comment by marshallford** [comment]
Interesting... I'm likely am running into this issue: https://github.com/istio/istio/issues/23392. Still digging.

**Comment by rytswd** [comment]
Most of the items raised here should be tackled in the v1.6.1 release.

### Fixed: Incorrect label handling
Custom labels are now correctly associated with the below fixes.
- v1.6.0 #22981
- v1.5.4 #23026

### Fixed: Duplicated Resources
When a gateway is given a custom name, ServiceAccount, Role, and RoleBindings were duplicated. This meant multiple gateways would share the same SA, etc. This has been changed to create custom SA, Role, RoleBinding based on the gateway name.
- v1.6.1 #23852 #24247
- merged to release-1.5 branch #23853 #24252

This made it difficult to use YAMLs generated by `istioctl manifest generate`, as the same exact resources are repeated in one file, which would cause error if you try to use it with Kustomize.

#### Remaining Items

Just for reference, the below items can still cause some unexpected behaviours.
As they are specific cases, I am closing this ticket. If any further discussion points, we should be able to create a separate ticket to track.

##### `multicluster` flag
Label is associated in created resources, but the resource name would default to use `istio-multicluster-ingressgateway`. Also, selectors only refer to `istio: ingressgateway` label which could have unexpected impact if you have multiple gateways.

##### `meshExpansion` flag
Label is not associated in the `meshexpansion-` resources (i.e. `gateway`, `vs-istiod`, and `dr-istiod`). This is probably not going to have any impact by itself even when you have multiple gateways. This, however, can cause some annoyance when combining `istioctl manifest generate` with `kustomize`, for instance.
